File: Data Structures & Algorithms/combination-target-sum/submission-1.go
// DP ends up generating and storing a bunch of intermediate combinations
// and keeps deep-copying them repeatedly, taking significant compute & space

func combinationSum(nums []int, target int) [][]int {
    // Sort now, so we may skip any values which are greater than target early on
    sort.Ints(nums)

    res := [][]int{}
    path := []int{}
    var dfs func(start, remainder int)
    dfs = func(start, remainder int) {
        if remainder == 0 {
            // mk a copy of path
            comb := make([]int, len(path))
            copy(comb, path)
            res = append(res, comb)
            return
        }
        for i := start; i < len(nums); i++ {
            x := nums[i]
            if x > remainder {
                // pruning early due to sorting, if not continue
                break
            }
            // choose
            path = append(path, x)
            // i, not i+1 : reuse allowed
            dfs(i, remainder-x)
            // un-choose (backtrack)
            path = path[:len(path)-1]
        }
    }
    dfs(0, target)
    return res
}

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res: List[List[int]] = []
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            if nums[i] > 0:
                break
            target = -nums[i]
            l, r = i+1, n-1
            while l < r:
                s = nums[l] + nums[r]
                if s == target:
                    res.append([nums[i], nums[l], nums[r]])
                    left_val, right_val = nums[l], nums[r]
                    while l < r and nums[l] == left_val:
                        l += 1
                    while l < r and nums[r] == right_val:
                        r -= 1
                elif s < target:
                    l += 1
                else:
                    r -= 1
        
        return res